### PR TITLE
introduce Formik and use it for the comment creation / editing form

### PR DESCRIPTION
**Pull request by aliceriot**
#### Pre-Flight checklist

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?

no ticket

#### What's this PR do?

introduces Formik and uses it to refactor out a `<CommentForm />` component, which lets us delete a bunch of yucky code. We were previously sharing from code from the predecessor of this component with the post editing form component - I'll swing back at some point to switch that over to Formik as well, but for now I'll leave it as-is.

#### How should this be manually tested?

Make sure you can make and edit comments, and reply to comments, without any problems. Editing a post should work as before as well.

### CODE REVIEW

File: static/js/components/CommentForm.js [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+              disabled={
+                isSubmitting || userIsAnonymous() || !values.commentText
+              }
+            >
+              Submit
+            </button>
+            {closeReply ? (
+              <button
+                onClick={preventDefaultAndInvoke(closeReply)}
+                className="cancel"
```
Comment (rhysyngsun) [inline static/js/components/CommentForm.js]: Should this be disabled when `isSubmitting` too?

File: static/js/components/CommentForm.js [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            }`
+          )
+        )
+      }
+    }
+  }
+})
+
+const CommentForm = connect<Props, OwnProps, _, DispatchProps, _, _>(
+  () => ({}),
```
Comment (rhysyngsun) [inline static/js/components/CommentForm.js]: I think you can just pass `null` here and avoid the extra func/call overhead.

File: static/js/components/CommentTree.js [hunk 3]
```diff
@@ -135,13 +155,8 @@ export default class CommentTree extends React.Component<Props> {
...
               beginEditing={e => {
-                if (beginEditing) {
-                  beginEditing(
-                    replyToCommentKey(comment),
-                    getCommentReplyInitialValue(comment),
-                    e
-                  )
-                }
+                e.preventDefault()
+                this.addCommentToState("replying", comment)
```
Comment (rhysyngsun) [inline static/js/components/CommentTree.js]: Nit: define constants for the state key

File: static/js/components/CommentTree.js [hunk 4]
```diff
@@ -90,14 +101,25 @@ export default class CommentTree extends React.Component<Props> {
     )
   }
 
+  removeCommentFromState = (key: string, comment: CommentInTree) => {
+    const newSet = new Set(this.state[key])
+    newSet.delete(comment.id)
+    this.setState({ [key]: newSet })
+  }
+
+  addCommentToState = (key: string, comment: CommentInTree) => {
```
Comment (rhysyngsun) [inline static/js/components/CommentTree.js]: Nit: Type `key` as `"replying" | "editing"`

File: static/js/components/CommentForm.js (hunk 1 above)
Comment (aliceriot) [reply static/js/components/CommentForm.js]: yeah, that makes sense I think

**Comment by aliceriot** [comment]
@rhysyngsun Ok those things should be addressed

**Pull request merged**